### PR TITLE
Parse GitHub URLs

### PR DESCRIPTION
**Pull request by jimhester**
A port of @jennybc's https://github.com/r-lib/remotes/pull/112 to
pkgdepends.

### CODE REVIEW

File: R/parse-remotes.R [hunk 1]
```diff
@@ -37,7 +37,7 @@ github_rx <- function() {
     ## Optional remote type
     "(?:github::)?",
     ## Username
-    "(?<username>(?:[^/]+))/",
+    "(?<username>(?:[[:alnum:]-]+))/",
```
Comment (jimhester) [inline R/parse-remotes.R]: This was needed so that the `github_rx()` did not match ssh Github URLs.

File: R/parse-remotes.R [hunk 2]
```diff
@@ -65,6 +93,7 @@ get_remote_types <- function(specs) {
 
   types[types == "" & grepl(standard_rx(), specs, perl = TRUE)] <- "standard"
   types[types == "" & grepl(github_rx(), specs, perl = TRUE)] <- "github"
+  types[types == "" & grepl(github_url_rx(), specs, perl = TRUE)] <- "github"
```
Comment (jimhester) [inline R/parse-remotes.R]: I don't think it makes sense to have a separate type for this, but it makes this code sort of weird.

File: R/type-github.R [hunk 3]
```diff
@@ -11,6 +11,10 @@
 parse_remote.remote_specs_github <- function(specs, config, ...) {
 
   pds <- re_match(specs, github_rx())
+  if (is.na(pds$.match)) {
```
Comment (jimhester) [inline R/type-github.R]: Not sure if there is a better way to do this, basically if the regular regex fails we try the url one.

File: R/type-github.R [hunk 4]
```diff
@@ -11,6 +11,10 @@
...
+    pds <- re_match(specs, github_url_rx())
+    pds$subdir <- ""
```
Comment (jimhester) [inline R/type-github.R]: The URL syntax does not currently support subdirectories, not sure if we want to add or not. For now I just set it to "" in this case.

File: R/parse-remotes.R (hunk 1 above)
Comment (gaborcsardi) [reply R/parse-remotes.R]: Hmmm. Do you know what characters are allowed in usernames?

File: R/parse-remotes.R (hunk 2 above)
Comment (gaborcsardi) [reply R/parse-remotes.R]: It's OK.

Comment (gaborcsardi) [reply R/parse-remotes.R]: Alternatively we could join the two regexes in a big `(...|...)`...

File: R/type-github.R (hunk 3 above)
Comment (gaborcsardi) [reply R/type-github.R]: I think it is good.

File: R/type-github.R (hunk 4 above)
Comment (gaborcsardi) [reply R/type-github.R]: OK.

**Comment by gaborcsardi** [comment]
THanks! Let's make sure that we capture all possible user names, otherwise it is fine.

**Comment by jimhester** [comment]
Ok I added a few cases with different username / repo names.

File: R/parse-remotes.R (hunk 1 above)
Comment (jimhester) [reply R/parse-remotes.R]: From https://www.npmjs.com/package/github-username-regex

- Github username may only contain alphanumeric characters or hyphens.
- Github username cannot have multiple consecutive hyphens.
- Github username cannot begin or end with a hyphen.
- Maximum is 39 characters

Repositories seem to have similar rules.

**Comment by gaborcsardi** [comment]
Thanks!

**Pull request merged**